0036-spsp-pull-payments/0036-spsp-pull-payments.md

```diff
@@ -0,0 +1,140 @@
+* **Pull Payment** - The process of pulling units of value from one account into another account.
+* **Pull Payment Agreement** - An agreement between the Payer and the Payee that specifies how much value can be pulled how often and for how long. 
+* **Token** - An opaque string that represents an identifier for the Pull Payment Agreement. It is part of the [Payment Pointer](../0026-payment-pointers/0026-payment-pointers.md) the SPSP Client needs to connect to the SPSP Server. 
+
+### Operation Overview
+
+The Payee's SPSP Client will set up a [STREAM](../0029-stream/0029-stream.md) connection to the Payer's SPSP Server using the Token, as described by the [Simple Payment Setup Protocol](../0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md). On connection, the SPSP Client tries to pull as much value from the SPSP server as possible given the Pull Payment Agreement.
+
+## Model of Operation
+
+### Negotiating the Pull Payment Agreement
+
+Prior to a Pull Payment, the Payer and the Payee have to negotiate a Pull Payment Agreement. This agreement SHOULD be comprised of the parameters defined in the `agreement`-object within the [Response Body](#Response-Body). The Payer's SPSP Server SHOULD store the negotiated Pull Payment Agreement parameters.
+
+The result of the negotiation is a Token representing the Pull Payment Agreement. The Token is required by the Payee's SPSP Client to connect to the Payer's SPSP server. It MUST be part of the [Payment Pointer](../0026-payment-pointers/0026-payment-pointers.md) and MUST therefore take any form that does not invalidate it.
+
+Implementations MAY try to parse the Token, however, it MUST support totally opaque tokens. 
+
+### Conducting the Pull Payment
+
+The Payee's SPSP Client opens a [STREAM](../0029-stream/0029-stream.md) connection to the Payer's SPSP Server as described in the [Simple Payment Setup Protocol](../0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md). Once this connection is established, the process continues as follows: 
+
+The SPSP Server begins sending ILP packets to complete the Pull Payment.
+  1. The SPSP Server will adjust their `sendMax` to reflect the amount they're willing to send.
+      * `sendMax` SHOULD be derived from the Pull Payment Agreement parameters.
+  2. The SPSP Client will adjust their `receiveMax` to reflect the amount they're willing to receive.
+      * `receiveMax` MAY be `Infinity`.
+  3. The SPSP Client's and Server's [STREAM Modules](../0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md#Definitions) will move as much value as possible while staying inside these bounds.
+  4. If the SPSP Server reaches their `sendMax`, they end the stream and the connection. If the SPSP Client reaches their `receiveMax`, they will end the stream and the connection.
```

Comment:
How does this behave when there's multiple open connections pulling?

Reply:
Pulling from the same endpoint? Should that be allowed?

Reply:
if it's not allowed we should say so. I think it would be nice if it were but I can't think of how we can make it work

Reply:
Should the server also only allow one stream at a time?

Reply:
I don't know if that works when the server is hosted by a wallet provider. Multiple entities may want to pull from different endpoints at the same time.

Reply:
It seems like the SPSP server (wallet provider or otherwise) should only allow a single connection and stream at a time to each pull payment pointer.

Reply:
Actually... after just trying to implement that limit, I think it is safe to have multiple connections and streams.
The SPSP server should only adjust their `sendMax` (step 1 above) for a single open stream at a time per pull payment pointer.
(over) simple example:
https://github.com/sabinebertram/ilp-spsp-pull-payment-server/commit/a58afe3470c926cb7c0ea1417e887e9cd76735ad

Reply:
Do you mean it is ok to have multiple connections and streams per server or per pull pointer?

Reply:
It is safe to have multiple connections and streams per pull pointer, but it SHOULD (MUST?) only set `sendMax` for one stream at a time per pull pointer.

Reply:
I would say MUST. I just don't see what we would gain by allowing multiple connections per pull pointer at a time.

Reply:
The pull server probably doesn't need to allow multiple connections or streams (unless maybe there's also data streaming happening), but I don't think this needs to comment on it since the basic concern is to not set `sendMax` on multiple streams at a time (for a single pull pointer).